pycyphal/presentation/subscription_synchronizer/transfer_id.py

```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2022 OpenCyphal
+# This software is distributed under the terms of the MIT License.
+# Author: Pavel Kirienko <pavel@opencyphal.org>
+
+from __future__ import annotations
+import asyncio
+import logging
+import typing
+from typing import Iterable, Any, Callable
+import pycyphal.presentation.subscription_synchronizer
+
+T = typing.TypeVar("T")
+_SG = pycyphal.presentation.subscription_synchronizer.SynchronizedGroup
+
+
+class TransferIDSynchronizer(pycyphal.presentation.subscription_synchronizer.Synchronizer):
+    """
+    Messages that share the same transfer-ID value are assumed synchronous.
```

Comment:
Shouldn't we make this use the sender ID as well? Even if it's not recommended, the only valid use of transfer IDs to group messages is when they are all from a single node that intentionally synchronizes the publishers for this purpose. Providing an implementation that doesn't have this constraint is a bad plan.

Reply:
Indeed, the grouping key should be a tuple of (sender node-ID, transfer-ID). Then subjects that are published by distinct nodes will never be grouped, which is desirable.

We could run separate matchers per sender node-ID also but the downside of this solution is that the memory requirement and worst-case complexity become dependent on the network configuration. For PyCyphal it makes no difference but if one attempted to implement this algorithm in a real-time system this would become a problem.